cmd/krew/cmd/list.go

```diff
@@ -24,11 +24,89 @@ import (
 
 	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/kubectl/pkg/util/slice"
 
+	"sigs.k8s.io/krew/pkg/constants"
 	"sigs.k8s.io/krew/pkg/installation"
 )
 
+type Entry struct {
+	Name    string
+	Version string
+}
+
+type Entries []Entry
+
+func sortByName(e Entries) Entries {
+	sort.Slice(e, func(a, b int) bool {
+		return e[a].Name < e[b].Name
+	})
+	return e
+}
+
+// Consume produces a junk GroupVersionKind for obj.GetObjectKind().GroupVersionKind().Empty() check to eat in PrintObj()
+type Consume struct {
+	APIVersion string
+	Kind       string
+}
+
+func (c Consume) SetGroupVersionKind(kind schema.GroupVersionKind) {
+	c.APIVersion, c.Kind = kind.ToAPIVersionAndKind()
+}
+
+func (c Consume) GroupVersionKind() schema.GroupVersionKind {
+	return schema.FromAPIVersionAndKind(c.APIVersion, c.Kind)
+}
+
+func (e Entries) GetObjectKind() schema.ObjectKind {
+	return Consume{constants.CurrentAPIVersion, constants.PluginKind}
+}
+
+func (e Entries) DeepCopyObject() runtime.Object {
+	return append(Entries{}, e...)
+}
+
+type ListFlags struct {
+	JSONYamlPrintFlags *genericclioptions.JSONYamlPrintFlags
+	OutputFormat       *string
+}
+
+func NewListFlags() *ListFlags {
+	outputFormat := ""
+	return &ListFlags{
+		JSONYamlPrintFlags: genericclioptions.NewJSONYamlPrintFlags(),
+		OutputFormat:       &outputFormat,
+	}
+}
+
+func (f *ListFlags) AllowedFormats() []string {
+	return f.JSONYamlPrintFlags.AllowedFormats()
+}
+
+func (f *ListFlags) ToPrinter() (printers.ResourcePrinter, error) {
+	outputFormat := *f.OutputFormat
+	var printer printers.ResourcePrinter
+
+	switch outputFormat {
+	case "json":
+		printer = &printers.JSONPrinter{}
+	case "yaml":
+		printer = &printers.YAMLPrinter{}
+	default:
+		return nil, genericclioptions.NoCompatiblePrinterError{OutputFormat: f.OutputFormat, AllowedFormats: f.AllowedFormats()}
+	}
+
+	return printer, nil
+}
```

Comment:
I'm still very uncomfortable that all this boilerplate code above this point is here in this file. IMO this is very easily testable and can also be reused by other modules. This file is for the `list` command, and should have a very high level abstraction. In contrast, the printing code has pretty low abstraction and does not fit well here.
As a result, you had to write a ton of integration tests for printing, because it was impossible to test the this independently of the `krew` machinery. This is not optimal.

Why do you think is this the best place for the printing details?

Reply:
> As a result, you had to write a ton of integration tests for printing, because it was impossible to test the this independently of the krew machinery

Printing is hardly my forte but wouldn't these integration tests need to exist regardless? Sorting should be covered (primarily) by a unit test but the integration test would implicitly verify that.

> Why do you think is this the best place for the printing details?

I do and I don't.

Ideally, much of the printing logic would be its own separate file `printer.go` that supports `krew info` as well (I made comments above semi-prototyping how that can go). Since that isn't done at this time, I just left the code in `list.go` to avoid confusing anyone into thinking its particularly re-usable. Would you like me to move it around?